### DISCUSSION TITLE
Submitting Multiple Records from a Repeater with a Single Button

### DISCUSSION THREAD

**Discussion by Sean-Roberts-89** [Budibase/budibase#10199] [Help]
Hello,

I am working on a project where I have a multistep form to record production data for a manufacturing process. The form consists of a work_order table, and after updating the work_order table, I need to create new entries in the reject_fact table based on the input provided by the user.

The reject_fact table has columns for the foreign keys work_order_id and defect_code_id, along with reject_units and reject_challenges. I have a data source that points to the defect_code table, which contains all possible reject codes. The goal is to have the user enter values for each reject code, and then submit all of these entries with a single "Submit All Reject Facts" button.

Currently, my setup includes a container with a button for submitting the reject_fact information, followed by a repeater that contains a new form for each reject code. The problem I'm facing is that the "Submit All Reject Facts" button cannot "see" the data inside the repeater, and I am unable to submit all the data at once. I tried placing the button inside the form, but since the form exists within the repeater, the button would create a record for each reject code one at a time, but I would like one button to submit all of the forms. 

When I set the bind_defect_code_id to {{ New Repeater.defect_code.defect_code_id }}, it works within the repeater but not in the button's action, as it shows "current handlebars syntax is invalid."

I am unsure how to proceed in order to have a single button submit all the data from the repeater to the reject_fact table. I would appreciate any guidance or suggestions on how to resolve this issue, or if there is a better way to achieve the desired functionality in Budibase.

Additionally, I am seeking a solution for situations where a user might mistakenly create more than one entry for the same reject code. The fact table should create a new row for each work order and reject code, but there should never be two of the same types of reject codes for the fact table for each work_order. In such cases, I would like to allow users to "edit" their previously entered data. All of the data should default to 0, so each time I click the button I would like to see it create ~20 new rows based on the data entered by the user, which should be 0 by default from the bindings. 

Thank you in advance!

![budibase](https://user-images.githubusercontent.com/125078607/229597236-aa3e842e-f020-495b-ba0c-fe8bce279cf5.JPG)

![schema](https://user-images.githubusercontent.com/125078607/229597493-1ac6e34a-186a-4809-a3c6-6e99e04beded.JPG)

**Comment by melohagan** [comment]
Hey @Sean-Roberts-89 

I think the challenge here is that currently only one form can be saved per **Save Row** action. In your case you want to save multiple dynamic form fields at the same time. 

This is currently not possible as the Form schema needs to be known up front. This issue here is related: https://github.com/Budibase/budibase/issues/5944

You may need to get users to enter and save facts one by one if that's an option? 

---

One possibility may be to use [App State](https://docs.budibase.com/docs/app-state) and keep track of the different field states through the [On Change event](https://docs.budibase.com/docs/actions#on-change), using the ID of the repeated row as the key.

Then your submit button could trigger an [App action](https://docs.budibase.com/docs/app-action) automation, passing the App State through as a field value, and then using [Automation looping](https://docs.budibase.com/docs/looping) to loop over each key in the dictionary and saving a row (or executing an insert/update custom query) for each item.

I appreciate that's a bit convoluted though. 

It's a different use-case, but there's a more detailed example here to guide you on using App State as a dictionary: https://github.com/Budibase/budibase/discussions/9536

**Comment by Sean-Roberts-89** [comment]
Thanks for the help! I tried this, by defining an action for each of the rejected fields, I set the key to rejectData, and the value is 

Set App StateKey: rejectDataValue: 

```js
{  
  ...current.rejectData,  
  [New Repeater.defect_code.defect_code_id]: {    
    work_order_id: Work Order Select.Fields.work_order,    
    defect_code_id: New Repeater.defect_code.defect_code_id,    
    reject_units: New Form.Fields.reject_units,    
    reject_challenges: New Form.Fields.reject_challenges  
  }
}
```


Then, I created a looping automation, with the binding as {{current.rejectData}}


I'm having difficulty understanding how to proceed. I'm not sure which JavaScript to use or if that's even the best approach, and I'm not very confident that I set the state correctly. The goal is to either create a new row in the database if it doesn't exist or update it if it does.
As far as having a button for each line, I might be able to do this as a last resort, but with around 25 lines, there's a risk of a user missing one of the button presses and making errors. If that's our only choice, then we'll have to do it.

Another possibility is to make changes to the database to simplify the process. The objective is to ensure that the user enters the quantity for each work order and the quantity for each defect code type for each work order. Would it be easier if this information was all on one 'fact' line that is just very wide and we pivot the information instead of creating multiple rows per transaction? I'm still trying to wrap my head around how we would do that given that I still do need an entry for each possible fact ID for each work order that the user enters.

**Reply by melohagan** [reply]
Hey @Sean-Roberts-89 I'm not sure if that JSON is pseudo-code or the actual value you have? 

Assuming a JavaScript binding, the key should really look more like ```[`$("New Repeater.defect_code.defect_code_id")`]: { ... }```

Or possibly:

```js
const id = $("New Repeater.defect_code.defect_code_id");
return {
  ...$("State.rejectData"),
  [`${id}`]: { ... }
}
```

I may need to see more screenshots, or if you wanted to send through an app export to support@budibase.com then I could take a closer look as well.

**Reply by Sean-Roberts-89** [reply]
That would be super helpful to me! I sent you an email @melohagan 

I appreciate you taking the time to look at it.

**Comment by Sean-Roberts-89** [comment]
@melohagan adding a tag :)

**Comment by melohagan** [comment]
Hey @Sean-Roberts-89 

The idea I was suggesting may be a little tricky as it involves JavaScript, but I'll point out a few things that aren't right:

You correctly have the *On Change* action updating state which is good! What's not right here is that you are trying to put JS inside a Handlebars binding.

<img width="1448" alt="Screenshot 2023-04-11 at 19 25 58" src="https://user-images.githubusercontent.com/101575380/231255517-b87a6903-786d-41bc-bde0-c387a4d0e05b.png">

Click on the JavaScript tab and **return** the new JSON app state:
<img width="868" alt="Screenshot 2023-04-11 at 19 28 52" src="https://user-images.githubusercontent.com/101575380/231256165-b30717e1-06f7-4067-bc46-56d072445a59.png">

The next part I was suggesting was to pass this through to the automation via the field:

<img width="498" alt="Screenshot 2023-04-11 at 19 30 40" src="https://user-images.githubusercontent.com/101575380/231256524-2d822eed-8e10-4815-afdd-149aafc9be70.png">

<img width="870" alt="Screenshot 2023-04-11 at 19 31 12" src="https://user-images.githubusercontent.com/101575380/231256647-5873e26e-0ce9-409e-8e78-5de179c9cbce.png">

I think for this particular use-case unfortunately you probably will struggle a little bit to build and maintain this if you're not familiar with JavaScript.

The Dynamic Form Fields implementation would be really useful here, and make this easier!

**Comment by Sean-Roberts-89** [comment]
Do you have any suggestions? I'm used to using R for work, but I'm open and
willing to learn. I also have chat GPT which can help me with some of the
javascript.

I need to do something with this app so that we can implement this, I would
hate to use Excel :(.

What would be ideal to put into the automation portion of the app?

Otherwise, I can't be the only person using this implementation, do you
know of a workaround?

**Comment by kpturner** [comment]
I have a repeater with a form in each, and a "save" button in each form to save the row.  
I also need a "catch all" button that will save all the rows in one fell swoop rather than clicking each save button.
Having read this thread, my intention is to create a button that executes some JavaScript to find and click all the "save" buttons in the repeater.  I will let you know how I get on.

Obviously the down side of this will be that I will be interrogating the DOM to get the buttons to click.  That means I will be relying on BB internals which may break when I upgrade to a new version.  It is a risk I need to take without a good safe "out of the box" solution for this.  

As a node.js engineer the JavaScript side of things is easy enough for me, but all this kinda defeats the object of using something like Budibase.  It is the tail wagging the dog.

My immediate stumbling block is that although the JS is running on the browser it does not have access to the DOM :(
